### PR TITLE
improvement: parse sql CREATE table names

### PR DESCRIPTION
**Pull request by mscolnick**
Fixes https://github.com/marimo-team/marimo/issues/2019

(thanks @akshayka for finishing this PR).

This adds "refs" and "defs" between SQL <-> SQL statements and Python -> SQL. (python leaks into sql's namepsace, but not vice-versa).

We define "definitions" as:
* `CREATE` for tables and views
* `ATTACH` for DBs

We draw references when:
* A `SELECT` contains a sql table, schema, or python df

Reference detection for schemas is incomplete but hopefully good enough for many use cases.

### CODE REVIEW

**Review by dmadisetti — COMMENTED** [review]
Cool! Closing #2019?

File: marimo/_ast/sql_visitor.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    def token_type(i: int) -> str:
+        return tokens[i][1]
+
+    # See https://duckdb.org/docs/sql/statements/create_table#syntax
+    # for more information on the CREATE TABLE syntax.
+    while i < len(tokens):
+        if (
+            keyword_token_str(i) == "create"
+            and token_type(i) == duckdb.token_type.keyword
```
[the remark this answers is not in the document]
Comment (mscolnick) [reply marimo/_ast/sql_visitor.py:137]: yea good call

File: marimo/_ast/sql_visitor.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                i += 1
+                if i < len(tokens) and keyword_token_str(i) == "if":
+                    i += 3  # Skip 'IF NOT EXISTS'
+                if i < len(tokens):
+                    table_name = token_str(i)
+                    # Remove quotes if present
+                    if table_name.startswith('"') and table_name.endswith('"'):
+                        table_name = table_name[1:-1]
+
+                    created_tables.append(table_name)
```
[the remark this answers is not in the document]
Comment (mscolnick) [reply marimo/_ast/sql_visitor.py:157]: i think its worth preserving the case in this and then check for case insensitive later, since you may want clearer error messages with the correct case

File: marimo/_ast/sql_visitor.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    sql_statement = remove_comments(sql_statement)
+
+    tokens = duckdb.tokenize(sql_statement)
+    created_tables: list[str] = []
+    i = 0
+
+    def token_str(i: int) -> str:
+        token = tokens[i]
+        start = token[0]
+        end = len(sql_statement) - 1
```
[the remark this answers is not in the document]
Comment (dmadisetti) [reply marimo/_ast/sql_visitor.py:121]: Oh, just saw the tests below with the weird table names. You might also be missing the `SELECT '''' AS s` edge case: https://duckdb.org/docs/sql/data_types/literal_types.html#escape-string-literals

and string consts `e'\n\b'`

```python
def token_str(i: int) -> str:
    token = tokens[i]
    start = token[0]
    boundary = r'[\s\-/]'
    quote = None
    if token_type(i) == duckdb.token_type.string_const:
        quote = sql_statement[start].lower()
        # Handle e'string' notation
        if quote == "e":
            start += 1
            # Should be '
            quote = sql_statement[start]
        # Quotes are released by doubling
        # From the duckdb docs:
        # SELECT '''' AS s
        # is valid (table name ')
        boundary = fr"{quote}(?=[^{quote}])"
        start += 1
    maybe_end = re.search(boundary, sql_statement[start:])
    end = start + maybe_end.start() if maybe_end else len(sql_statement)
    if i + 1 < len(tokens):
        # For tokens squashed together e.g. '(select' or 'x);;;;'
        # in (select * from x);;;;
        end = min(end, tokens[i + 1][0]) 
    if quote:
        return re.sub(quote * 2, quote, sql_statement[start:end])
    return sql_statement[start:end]
 ```
 
which also strips quotes for free

[the remark this answers is not in the document]
Comment (mscolnick) [reply marimo/_ast/sql_visitor.py:121]: this doesnt handle the case:
```sql
       CREATE TABLE "my--table" (
            "column/*with*/comment" INT,
            "another--column" VARCHAR
        );

        CREATE TABLE my_table_with_select AS
        SELECT *
        FROM (
            VALUES
            ('a', 1),
            ('b', 2)
        ) AS t("col--1", "col--2");

        CREATE TABLE "my/*weird*/table" (id INT);

        CREATE TABLE "with a space" (id INT);
```

[the remark this answers is not in the document]
Comment (mscolnick) [reply marimo/_ast/sql_visitor.py:121]: i added:
```python
if sql_statement[start] == '"':
            # If it starts with a quote, find the matching end quote
            end = sql_statement.find('"', start + 1) + 1
```

[the remark this answers is not in the document]
Comment (dmadisetti) [reply marimo/_ast/sql_visitor.py]: Oh woops, I broke it when checking for the e-strings. This still doesn't get single quotes or escaped quotes

fixed my bug, string table names are still "identifiers" not string_const unless they have the `e`

```python
def token_str(i: int) -> str:
    token = tokens[i]
    start = token[0]
    boundary = r'([\s\-/])'
    quote = None
    if token_type(i) == duckdb.token_type.string_const:
        # Handle e'string' notation
        if sql_statement[start].lower() == "e":
            start += 1
    if sql_statement[start] in ("\"", "'"):
        quote = sql_statement[start]
        # Quotes are released by doubling
        # From the duckdb docs:
        # SELECT '''' AS s
        # is valid (table name ')
        boundary = fr"(?<!{quote})(?:{quote * 2})?({quote})(?!{quote})"
        start += 1
    maybe_end = re.search(boundary, sql_statement[start:])
    end = start + maybe_end.end() - 1 if maybe_end else len(sql_statement)
    if i + 1 < len(tokens):
        # For tokens squashed together e.g. '(select' or 'x);;;;'
        # in (select * from x);;;;
        end = min(end, tokens[i + 1][0])
    if quote:
        return re.sub(quote * 2, quote, sql_statement[start:end])
    return sql_statement[start:end]
```

Which seems to work: https://marimo.app/l/tc7cqv

**Review by akshayka — APPROVED** [review]

File: marimo/_ast/sql_visitor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmadisetti) [reply marimo/_ast/sql_visitor.py]: cc @akshayka I think the current logic will still break for the escaped string literal notation https://duckdb.org/docs/sql/data_types/literal_types.html#escape-string-literals
